File: public/core/src/main/java/se/infomaker/frt/moduleinterface/action/module/ModuleActionHandler.java
package se.infomaker.frt.moduleinterface.action.module;

import android.content.Context;
import android.os.Bundle;

import java.util.Iterator;

import se.infomaker.frt.moduleinterface.action.ActionHandler;
import se.infomaker.iap.action.Operation;
import timber.log.Timber;

import static se.infomaker.frt.moduleinterface.action.module.Module.MODULE_ID;

public class ModuleActionHandler implements ActionHandler {
    public static final String OPEN_MODULE = "openModule";

    @Override
    public String perform(Context context, Operation operation) {
        Timber.d("ModuleActionHandler OpenModule Operation: %s", operation);
        switch (operation.getAction()) {
            case OPEN_MODULE: {
                Bundle bundle = new Bundle();

                Iterator<String> iterator = operation.getParameters().keys();
                while(iterator.hasNext()) {
                    String key = iterator.next();
                    bundle.putString(key, operation.getParameter(key));
                }
                // Module is defined by id externally but moduleId internally -> remapping
                if (!bundle.containsKey(MODULE_ID)) {
                    bundle.putString(MODULE_ID, bundle.getString("id"));
                }
                try {
                    Module.open(context, operation.getParameter(ModuleActivity.MODULE_NAME), bundle);
                } catch (InvalidModuleException e) {
                    Timber.e(e, "Trying to open invalid module");
                    return null;
                }
                return "";
            }
            default: {
                return null;
            }
        }
    }

    @Override
    public boolean canPerform(Context context, Operation operation) {
        return OPEN_MODULE.equals(operation.getAction()) && Module.isValid(operation.getParameter(ModuleActivity.MODULE_NAME));
    }
}

File: public/core/src/main/java/se/infomaker/iap/action/module/ModuleActionHandler.kt
package se.infomaker.iap.action.module

import android.content.Context
import se.infomaker.frt.moduleinterface.action.module.Module.MODULE_ID
import se.infomaker.iap.action.ActionHandler
import se.infomaker.iap.action.Operation
import se.infomaker.iap.action.Result
import timber.log.Timber

object ModuleActionHandler : ActionHandler {
    override fun canPerform(context: Context, operation: Operation): Boolean = Module.isValid(operation.getParameter(ModuleActivity.MODULE_NAME) ?: "")

    override fun isLongRunning(): Boolean = false

    override fun perform(context: Context, operation: Operation, onResult: Function1<Result, Unit>) {
        val bundle = operation.parametersAsBundle()
        // Module is defined by id externally but moduleId internally -> remapping
        if (!bundle.containsKey(MODULE_ID)) {
            bundle.putString(MODULE_ID, bundle.getString("id"))
        }
        try {
            operation.getParameter(ModuleActivity.MODULE_NAME)?.let { name ->
                if (name.isNullOrEmpty()) {
                    onResult.invoke(Result(false, operation.values, "No module name"))
                } else {
                    Module.open(context, name, bundle)
                    onResult.invoke(Result(true, operation.values, null))
                }
            }

        } catch (e: InvalidModuleException) {
            Timber.e(e, "Trying to open invalid module")
            onResult.invoke(Result(false, operation.values, "Trying to open invalid module"))
        }
    }
}
